KOGITO-4724 SQL database scripts bundle

uteegozi opened the pull request:
<details>
<summary>
Bundle db scripts into a single .zip file per repository. 

Relevant repositories: kogito-runtimes, kogito-apps
I will open a similar PR from kogito-apps once this PR was built ok.

The kogito-apps repo creates in addition to it's repos .zip also the **kogito-all** zip file merging the kogito-runtimes and kogito-apps created .zip artifacts 
</summary>

**JIRA issue:**
https://issues.redhat.com/browse/KOGITO-4724

**Related PR request:**
"KOGITO-4724 SQL database scripts bundle" in kogito-apps

**Comments**
The implementation answers to the requirements of the JIRA: zip per project, scripts inside zip aggregated per db, same groupId for this artifact for all repositories; 
The chosen solution is a compromise between a) a hard-coded path to each module containing *.sql files (does allow for more concise .zip structure but every new/change in module with *.sql resources requires update of assembly definitions) and b) completely open aggregation collecting any *.sql file there is (no updates of assembly definitions needed but also no order by DB inside .zip). 
The chosen solution does aggregate per DB with full path structure in .zip like in the project; addition of new/changes in modules of an already defined DB (right now postgresql and jdbc) does not require any updates in the assembly descriptor; addition of resources for a new DB (e.g. MariaDB), that should be aggregated, does

_the resulting assembly structure is:_
"db name" / "path inside module containing scripts"
e.g. /postgresql/addons/persistence/postgresql-persistence-addon/src/main/resources/sql/create_tables.sql

_the resulting maven repo file structure is:_
org -> kie -> kogito -> persistence -> ddl (groupId)
   -> kogito-runtimes (artifactId) -> "version" -> kogito-runtimes-"version"-db-scripts.zip
   -> kogito-apps (artifactId) -> "version" -> kogito-apps-"version"-db-scripts.zip
   -> kogito-all (artifactId) -> "version" -> kogito-all-"version"-db-scripts.zip

**Important to know:**
If at one point folder the project structure changes and therefore the assembler won't find any fitting files, an error will be thrown by the plugin which kills the execution of the build. 

</details>

uteegozi commented:
Hi, I wanted to check the logs but link for "Please check the logs" or "Details" next to Linux seem to be incorrect:
https://eng-jenkins-csb-business-automation.apps.ocp4.prod.psi.redhat.com/job/KIE/job/kogito/job/pullrequest/job/kogito-runtimes/266/display/redirect

uteegozi commented:
jenkins retest this

cristianonicolai commented:
jenkins please retest this

uteegozi commented:
Problem: Timeout has been exceeded
trying 
jenkins please retest this
one more time

cristianonicolai commented:
jenkins please retest this

uteegozi commented:
Problem: Timeout has been exceeded

Logs show the following error (same error like in the previous builds):
[2021-05-26 03:07:12,373] ERROR Failed to submit metrics to Kafka topic __confluent.support.metrics (due to exception): java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.TimeoutException: Topic __confluent.support.metrics not present in metadata after 10000 ms. (io.confluent.support.metrics.submitters.KafkaSubmitter)

After the error something wakes up every 5 minutes for about 3 hours until finally an interrupt is sent: 
****
Cancelling nested steps due to timeout
Sending interrupt signal to process
***

Please advise how to proceed.

cristianonicolai commented:
jenkins please retest this

uteegozi commented:
got to build-apps stage this time - try to build again?

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented:
jenkins please retest this

uteegozi commented:
checked linux logs and fixed error - jenkins please retest this

cristianonicolai commented:
jenkins please retest this

MarianMacik commented on kogito-build/kogito-ddl/pom.xml:
```diff
@@ -0,0 +1,42 @@
+
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>org.kie.kogito.persistence.ddl</groupId>
+    <artifactId>kogito-runtimes</artifactId>
```
`kogito-runtimes` is already an `artifactId` of the top-level repository POM. It's a good convention to have a module in the same directory as its `artifactId`, so we can name this module `kogito-ddl` and use `finalName` configuration tag to change the name of the resulting zip file to `kogito-runtimes` wdyt?

uteegozi replied:
just making sure I understand correctly: with "to have a module in the same directory as its artifactId," you mean have the module name match the artifactId?

MarianMacik replied:
@uteegozi Correct.

uteegozi replied:
@MarianMacik
Right, so I would like then to remind everybody that we are going to have **three** new artifacts connected to db scripts: 
* one is build in the kogito-runtimes project only for the runtime scripts; 
* one will be built in the kogito-apps project (PR shortly) only for the apps scripts 
* and the third will repackage those two earlier artifacts into a final script zip also inside the apps project (which is possible since **apps** is always built after **runtimes** in all relevant pipelines).

So, while your suggestion (having a module kogito-ddl and rename the artifact) would be fine if we had just this one assembly, I don't think it's fitting the current situation. 
I did play with defining a `kogito-ddl `module in both projects: .zip assemblies can then be installed under that same artifact easily with different names, it's just that the .pom files from both modules have the same names and the last copied overrides the first. I also have the third repackaged .zip which needs to be build in a separate module (tried to build apps scripts and the repackaged script in one module with two descriptors - didn't work because of the dependencies for repackaging) and as such would always be under a different path than the other two anyway

So my solution suggestion is very much like what I suggested in my initial PR commit comment but taking module name to artifact id convention into account as well as not repeating existing module names:
_the resulting maven repo file structure is:_
org -> kie -> kogito -> persistence -> ddl (groupId)
-> kogito-runtimes-ddl (artifactId, module name) -> "version" -> kogito-runtimes-ddl-"version"-db-scripts.zip
-> kogito-apps-ddl (artifactId, module name) -> "version" -> kogito-apps-ddl-"version"-db-scripts.zip
-> kogito-ddl (artifactId, module name) -> "version" -> kogito-ddl-"version"-db-scripts.zip
with `kogito-ddl `artifact containing it all.

Thank you for taking the time and what do you say?

Btw, `finalName` does change the build name only inside the `target` folder, not when being installed in repo

MarianMacik replied:
Correct. I originally thought that the groupIds are different. Moreover, I didn't realize `finalName` is not taken into account when installing to a local repository. Looks good now. Thanks!

MarianMacik commented on kogito-build/kogito-ddl/src/assembly/db-scripts.xml:
```diff
@@ -0,0 +1,28 @@
+            <directory>${project.root.dir}</directory>
+            <outputDirectory>postgresql</outputDirectory>
+            <includes>
+                <include>**/*postgresql*/src/main/resources/**/*.sql</include>
```
This results in a very long path (e.g. /postgresql/addons/persistence/postgresql-persistence-addon/src/main/resources/sql/create_tables.sql) in the zip file. Wouldn't it be better to rather specify the full directory and here only include `*.sql` pattern? This would make the following hierarchy:

```
postgresql
-- create_tables.sql
-- exists_tables.sql
jdbc
-- create_tables.sql
-- exists_tables.sql
```

I know that this will make assembly file less generic, covering only specific directories, but even now we would have to edit it in case a new DB is added, so I don't think it is an issue.

uteegozi replied:
Yes, I agree it's a very long path and if the consensus is to go for better readability of the .zip file than for a more general assembly process and I am happy to change that.
Please keep in mind that we are assembling 3 .zip files - one inside kogito-runtimes for all .sqls found there, a similar one inside the kogito-apps for all .sql there and then the "full" .zip assembled also inside the kogito-apps build process by repackaging the just created "per project" .zips. Need to make sure that .sql file names are never the same for a specific DB, even between projects.

MarianMacik replied:
As long as we keep the scripts named uniquely, it should be fine.

MarianMacik reviewed (commented):
Just 2 possible enhancements, let me know what you think.

tiagodolphine commented on kogito-build/kogito-runtimes-ddl/pom.xml:
```diff
@@ -11,8 +11,8 @@
 
     <modelVersion>4.0.0</modelVersion>
     <groupId>org.kie.kogito.persistence.ddl</groupId>
```
Shouldn't the groupId be just `org.kie.kogito` following the same as other modules?

uteegozi replied:
@cristianonicolai Cristiano suggested the groupid. we are also having a module on the kogito-apps project being built under the same groupId, so both ddl artifacts from the two projects would have a common location inside the repo:
 org.kie.kogito.persistence.ddl -> kogito-runtimes-ddl
and 
 org.kie.kogito.persistence.ddl -> kogito-ddl  (containing both runtmes and apps ddls)

cristianonicolai replied:
yeah, I was thinking to have it as a subgroup just to make it easier to find the artifacts. `org.kie.kogito.ddl` could to too.

uteegozi replied:
@tiagodolphine , @cristianonicolai - whatever you decide - for what it's worth I would go with the current standard.

uteegozi replied:
so change groupid to org.kie.kogito?

cristianonicolai commented:
Jenkins rerun examples tests

cristianonicolai reviewed (commented):
@uteegozi we should probably move this new module to `https://github.com/kiegroup/kogito-runtimes/tree/master/addons/persistence` wdyt?

uteegozi commented:
> @uteegozi we should probably move this new module to `https://github.com/kiegroup/kogito-runtimes/tree/master/addons/persistence` wdyt?

@cristianonicolai I thought the current location is good being like some module spanning location for project wide definitions, explanations and actions. But addons/persistence seems also to be a logical fit. Anyways, as another Jira is right now changing the structure of the addons folder and the addons/persistence location is changing, maybe you want to move it later once that is final?

cristianonicolai commented:
@uteegozi sure, we can wait for that PR to be merged with the new structure

cristianonicolai commented:
@uteegozi restructure is now merged

uteegozi commented:
changed groupid to standard org.kie.kogito
changed artifact name to kogtio-ddl-runtimes
moved module under addons/common/persistence under new name "ddl" (in conformance with the new addon module names)

tiagodolphine commented:
Jenkins retest this

cristianonicolai commented:
Jenkins rerun apps tests

cristianonicolai commented:
jenkins please retest this

cristianonicolai commented:
Jenkins rerun runtimes tests

cristianonicolai commented:
Jenkins rerun apps tests
